index.js

```diff
@@ -3656,6 +3656,42 @@
       [Fn(a, b), m(a), m(a)],
       Z.sortBy);
 
+  //# zipWith :: (Applicative m, Foldable m, Monoid m) => (a -> b -> c) -> m a -> m b -> Arrray c
+  //.
+  //. Creates a new array out of the two supplied structures
+  //. by applying the given function to each equally-positioned
+  //. pair of elements from each given [Foldable][]. The returning
+  //. array is truncated to the length of the shorter of the two
+  //. input Foldables.
+  //.
+  //. ```javascript
+  //. > S.zipWith(S.add, [1, 2, 3], [3, 2, 1])
+  //. [4, 4, 4]
+  //.
+  //. > S.zipWith(a => b => [a, b], [1, 3, 5], [2, 4, 6])
+  //. [[1, 2], [3, 4], [5, 6]]
+  //.
+  //. > S.zipWith(a => b => Cons(a, Cons(b, Nil)), Cons('foo', Nil), Cons('bar', Nil))
+  //. [Cons('foo', Cons('bar', Nil))]
+  //. ```
+  function zipWith(f, a, b) {
+    var idy = 0;
+    var aIsShorter = S.size(a) < S.size(b);
```

Comment:
This is inefficient because `S.size` traverses the whole structure to determine its size (setting aside the special case for arrays). Additionally, we avoid using `S` functions internally to avoid additional type checking.

A good starting point would be to convert both structures to arrays:

```javascript
//  toArray :: Foldable f => f a -> Array a
function toArray(foldable) {
  return Z.reduce(function(xs, x) { xs.push(x); return xs; }, [], foldable);
}
```

```javascript
function zipWith(f, xs, ys) {
  var $xs = toArray(xs);
  var $ys = toArray(ys);
  ...
}
```

To avoid using `aIsShorter ? ... : ...` in four places (including twice in a reducing function), we could truncate the longer array. To avoid allocating two new arrays, we could update the `length` properties like so:

```javascript
function zipWith(f, xs, ys) {
  var $xs = toArray(xs);
  var $ys = toArray(ys);
  $xs.length = $ys.length = Math.min($xs.length, $ys.length);
  ...
}
```

From this point the code for combining `$xs` and `$ys` should be relatively simple. :)

Reply:
Gotcha! I didn’t know about to array!!! That’s really nice!

Reply:
Wait, toArray doesn’t exist! 😱 I’ll define it as you put it 🤔

Reply:
Should we expose toArray?

Reply:
One day, perhaps, but let's stay focused on the task at hand. Define `toArray` near the top of the file with the other private functions. ;)